### ISSUE TITLE
Markers no removed

### ISSUE THREAD

**Issue by VictorPulzz** [react-native-maps/react-native-maps#4214]
# Bug report

## Summary
I use last version react-native-maps, I add custom marker on map, using array, but when I remove marker, marker not removed from map...

## Video
Just check video
https://user-images.githubusercontent.com/23406402/165700618-c313e746-db29-4882-a87e-e792b865b027.mp4

## Environment info
```bash
"react": "17.0.2",
"react-native": "0.67.4",
"react-native-maps": "^0.31.0",

Simulator Target Platform: iOS - 15.2.1 iPhone 11 (using Apple Maps)
Simulator Target Platform: iOS - 13.5.1 iPhone 6 (using Apple Maps)
```

```bash
System:
    OS: macOS 12.2.1
    CPU: (8) x64 Apple M1
    Memory: 35.43 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.13.1 - ~/.nvm/versions/node/v16.13.1/bin/node
    Yarn: 1.22.17 - ~/.nvm/versions/node/v16.13.1/bin/yarn
    npm: 8.1.2 - ~/.nvm/versions/node/v16.13.1/bin/npm
    Watchman: 2022.03.07.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.2, iOS 15.2, macOS 12.1, tvOS 15.2, watchOS 8.3
    Android SDK:
      API Levels: 30, 31, 32
      Build Tools: 30.0.2, 30.0.3, 32.0.0, 32.1.0
      System Images: android-30 | ARM 64 v8a, android-30 | Intel x86 Atom_64, android-30 | Google APIs ARM 64 v8a, android-30 | Google APIs Intel x86 Atom, android-30 | Google APIs Intel x86 Atom_64, android-30 | Google Play ARM 64 v8a, android-31 | Google APIs ARM 64 v8a, android-32 | Google APIs ARM 64 v8a
      Android NDK: Not Found
  IDEs:
    Android Studio: 2021.1 AI-211.7628.21.2111.8193401
    Xcode: 13.2.1/13C100 - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.11 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2 
    react-native: 0.67.4 => 0.67.4 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

## Steps to reproduce
1. Follow instruction steps for this package
2. Add a marker by tapping anywhere in the MapView
3. Tap 'clear markers' to set the marker array to empty

Describe what you expected to happen:
1. Markers was remove

Describe what actually happens:
1. Markers not removed

## Reproducible sample code
```js
const Map = () => {
const [points, setPoints] = useState([]);

  const addPoint = useCallback(
    coords => {
      setPoints(prevState => {
        return [
          ...prevState,
          {
            ...coords,
          },
        ];
      });
    },
    [],
  );

  const removePoint = useCallback(() => {
    setPoints(prevState => [...prevState.slice(0, -1)]);
  }, []);

 return (
      <View>
            <MapView
                camera={{
                   center: {
                      latitude: 54.03826514370414,
                     longitude: 82.96180595441342,
                   }
                  pitch: 0,
                  heading: 0,
                  altitude: 500,
                  zoom: 12,
               }}
               style={{
                 position: 'absolute',
                 top: 0,
                 bottom: 0,
                 right: 0,
                 left: 0,
               }}
          onPress={e => addPoint(e.nativeEvent.coordinate)}
        >
          {points.map((item, index) => (
            <Marker key={index} coordinate={item} />
          ))}
        </MapView>
       <TouchableOpacity
                onPress={removePoint}
                style={{position: 'absolute', bottom: 50, zIndex: 100}}
      >Remove point</TouchableOpacity>
</View>
)
}
```

**Comment by l-barbosa** [comment]
@VictorPulzz I've been experiencing this issue for quite a while now. 
Just like you, I saw issue #4102 and that it wasn't reproducible and the poster ended up giving up on it, only to try it later and it did work.

I've been working on other sections of my app (that hasn't been released yet) and was hoping for this to be magically fixed by a new release but 0.31.0 doesn't fix it (I know... why would it if the issue hasn't been confirmed yet? wishful thinking...).
As my app is getting close to being ready and this being one of the last outstanding issues I decided to dedicate some more time to it. I was happy to see that I'm not alone here.

Since I can't share my app's code, I put together a brand new project today to demonstrate my use case as close as possible. And... surprise surprise... just like the poster of #4102, it worked fine on my new project...
Also tried your sample code on that new project (had to make some minor changes to get it to work) and it worked fine as well.

In my use case, I'm showing the markers conditionally based on the zoom level (delta). When the user zooms in after a certain level, the markers appear. The user zooms out, markers disappear.

Not being familiar with the internals of this project, here's what I've found in my attempts to work around this:
- It's not related to the custom view/marker as I initially thought (and so did the poster of #4102). If I replace them with an image or just a plain default marker, they aren't being removed when I zoom out. What tricks you into thinking that it's related to the custom markers, is that the custom markers DO get removed when you want them but in their place they leave the standard marker, that no longer disappears from that point forward, staying even when the custom marker reappears when it is supposed to.
- It's not specific to Apple Maps or Google Maps. Even though I'm on iOS, I was using Google Maps. I switched to the default provider and the issue still exists.

I can still upload the demo project but I don't see any benefit in doing so since it's working just fine.

Anyone more familiar with the internals of the project (@Simon-TechForm?) has any suggestion as to what I could test next, either on the project where it is failing, or on the demo project in order to get it to fail?
Could the view structure, navigation and all that stuff inherent to a proper project impact this in a way that a demo project would not? It's strange that in both cases we're failing to replicate the issue in a more basic demo project.

Any help/hint is appreciated!

**Comment by l-barbosa** [comment]
Some more details.
I went into `AirMap.m` and tried to log data as views get removed:

```
- (void)removeReactSubview:(id<RCTComponent>)subview {
    // similarly, when the children are being removed we have to do the appropriate
    // underlying mapview action here.
    NSLog(@"Removing view %@", [subview class]);
    if ([subview isKindOfClass:[AIRMapMarker class]]) {
        [self removeAnnotation:(id<MKAnnotation>)subview];
        NSLog(@"Removing annotation, total annotations %ld", [self annotations].count);
   }
...
}
```

When I zoomed out and the custom markers where removed but in their place the default marker appeared, this got logged:

```
Removing view AIRMapMarker
Removing annotation, total annotations 102
Removing view AIRMapMarker
Removing annotation, total annotations 101
...
Removing view AIRMapMarker
Removing annotation, total annotations 1
```

This didn't seem to make sense since I was still seeing the markers/annotations.
I wondered if something else was happening async so I added this to the end of the `removeReactSubview`:

```
dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 3 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        NSLog(@"Number of annotations now after delay: %ld", [self annotations].count);
    });
```
and the output this time was:

```
Removing view AIRMapMarker
Removing annotation, total annotations 101
....
Removing view AIRMapMarker
Removing annotation, total annotations 1
Number of annotations now after delay: 101
```

So they seem to be getting removed but are added again at some point.
A breakpoint at `-(void)addSubview:(UIView *)view` shows it being called in the context of `- (void)didUpdateReactSubviews` which took me back to a possible issue with my react code but when I finally added logging to:
 `-(void)addSubview:(UIView *)view`

> `NSLog(@"insertReactSubview called (%ld|%ld)", [self annotations].count, [self reactSubviews].count);`

`- (void)insertReactSubview:(id<RCTComponent>)subview atIndex:(NSInteger)atIndex`
> `NSLog(@"insertReactSubview called (%ld|%ld)", [self annotations].count, [self reactSubviews].count);`

`- (void)removeReactSubview:(id<RCTComponent>)subview`
> `NSLog(@"removeReactSubview (%ld|%ld)", [self annotations].count, [self reactSubviews].count);`

AND on the react side on `MapView.js` right before the return of the `render` method:
> `console.log("react render, number of children of map: ", props.children.length)`

Here's a sample of what I received, when moving from a zoomed in view (should have markers) to a zoomed out view (should not have markers): 
(reduced the total number of markers to 5 to keep things sane...)

```
[javascript] 'react render, number of children of map: ', 5
[javascript] 'react render, number of children of map: ', 5
[javascript] 'react render, number of children of map: ', undefined
removeReactSubview (5|4)
insertReactSubview called (6|5)
removeReactSubview (5|4)
removeReactSubview (4|3)
insertReactSubview called (5|4)
insertReactSubview called (6|5)
removeReactSubview (5|4)
removeReactSubview (4|3)
removeReactSubview (3|2)
insertReactSubview called (4|3)
insertReactSubview called (5|4)
insertReactSubview called (6|5)
removeReactSubview (5|4)
removeReactSubview (4|3)
removeReactSubview (3|2)
removeReactSubview (2|1)
insertReactSubview called (3|2)
insertReactSubview called (4|3)
insertReactSubview called (5|4)
insertReactSubview called (6|5)
removeReactSubview (5|4)
removeReactSubview (4|3)
removeReactSubview (3|2)
removeReactSubview (2|1)
removeReactSubview (1|0)
insertReactSubview called (2|1)
insertReactSubview called (3|2)
insertReactSubview called (4|3)
insertReactSubview called (5|4)
insertReactSubview called (6|5)
addSubview called: (6 | 5)
addSubview called: (6 | 5)
addSubview called: (6 | 5)
addSubview called: (6 | 5)
addSubview called: (6 | 5)
```

So, on the react side, things seem to be working as expected. We move from a render of 5 children to none which seems to rule out a mistake on my react code.
On the iOS side, something strange is happening. I kept the whole log because the cadence of calls may help understand what's happening. It seems to be snowballing to some degree. When I had 100 markers this was even clearer.

One other thing to add is that as I zoom in and out of the threshold that should trigger markers to appear/disappear, my custom markers appear and disappear as expected, but the default markers just keep adding up, and the total number of annotations logged keeps increasing (5, 10, 15, 20, ...)

I'm hoping this can help understand where to look for the issue.

**Comment by iarmankhan** [comment]
Same issue when using Children in Marker component

**Comment by l-barbosa** [comment]
I ran my app on Android and the issue did not manifest there.
I'll be focusing my efforts on iOS and I'll post any updates.

**Comment by iarmankhan** [comment]
Did you find any solutions for IOS? I tried updating the library to latest version but it's still happening for me. 
@l-barbosa Do we not have any workarounds for this?

**Comment by l-barbosa** [comment]
No, I'm not a contributor to this project so I'm not that familiar with it, was just using the library and ran into this issue.
I'm trying to either put together a repository where the issue can be easily reproduced or narrow down the source of the issue. Whatever I can do to make it easier for regular maintainers.
I'll post any update on this thread.

**Comment by iarmankhan** [comment]
Thanks, from another thread a workaround could be to give key as markers.length to Map component, this will forcefully re-render it. I couldn't try this out due to some other issues but maybe it can work for you. @l-barbosa

**Comment by l-barbosa** [comment]
By setting up a few breakpoints as the views are added/removed and following the execution I found that it was [React Native Reanimated](https://github.com/software-mansion/react-native-reanimated) that was adding views back as they were removed.

To test this I went back to the project I set up to demonstrate my issue and that initially was working fine and added reanimated to it.
As soon as I added a layout animation similar to the one I'm using in my project, the issue started manifesting itself.
You can check out the demo project [here](https://github.com/l-barbosa/react-native-maps-reanimated-issue).

I don't know if this was the source of the issue for other posters but I'm confident that it is what's causing the issue in my case.
It's not clear if the issue is with the way `react-native-maps` handles the child views or if it's `react-native-reanimated` that's doing something that it's not supposed to.
I've posted the issue [there too](https://github.com/software-mansion/react-native-reanimated/issues/3210).

**Comment by iarmankhan** [comment]
This is really frustrating that none of the maintainer is even replying on this issue, I've tried everything but nothing can make it work properly in my existing project.

**Comment by l-barbosa** [comment]
It's never easy when an issue is caused by the interaction between two libraries...
I will probably just remove the use of the `reanimated` library on that particular screen, replacing it with standard React animations. 
If your issue is the same as mine and you do the same, you should be able to get around the issue.

**Comment by iarmankhan** [comment]
Actually no I'm not even using animations, react-native-reanimated library is there due to react-navigation. For now what I did was to add key prop to map, which is really not good as it reloads the whole map and it flickers. But that's easier than rewriting whole app.

**Comment by l-barbosa** [comment]
Sorry to hear that we're not dealing with the same case.

In case anyone is experiencing the same specific issue I was able to work around it by patching `react-native-reanimated`.
You can find [the patch](https://github.com/l-barbosa/react-native-maps-reanimated-issue/blob/master/patches/react-native-reanimated%2B2.8.0.patch) in the demo repository.

**Comment by PaulOskarSoe** [comment]
@l-barbosa Your workaround works nicely and hopefully reanimated 2 maintainers will work on children excluding logic. Thanks for the patch!!

**Comment by xanderdeseyn** [comment]
Thank you for the detailed write-up and fix @l-barbosa !

**Comment by Simon-TechForm** [comment]
Huge props to @l-barbosa. This explains the difficulty trying to reproduce this. I will leave the issue open looking forward to the answer from react-native-reanimated.

**Comment by Simon-TechForm** [comment]
Hello everyone. I've just subscribed to the issue created by @l-barbosa (thanks again) in `react-native-reanimated`, and will close this for now. If it turns out that there's anything that needs to be done on our side, I will re-open. :)

**Comment by RamonAmaro** [comment]
para quem teve este problema, é só aplicar esse patch no pacote do react-native-reanimated, aqui corrigiu o problema.

https://github.com/l-barbosa/react-native-maps-reanimated-issue/blob/master/patches/react-native-reanimated%2B2.8.0.patch

**Comment by vxm5091** [comment]
@l-barbosa is the patch still working on your end? I applied it to `react-native-reanimated 2.8.0` and still am randomly getting this issue. I'm using reanimated animations a decent amount throughout the app but otherwise is there something I might be missing? 

Thanks a lot for sharing the patch and opening these issues! 

![IMG_9B3AE1F9541E-1](https://user-images.githubusercontent.com/57078583/200075941-0fa2a31f-3935-4a98-91da-9aa9f8cb1cae.jpeg)

**Comment by PaulOskarSoe** [comment]
@vxm5091 I have been using this patch for 5 months and it fixed the issue for me.

**Comment by vxm5091** [comment]
Yep, got it working on my end as well. Upgrading to the latest version of `react-native-maps` + the patch did the trick!

**Comment by l-barbosa** [comment]
@vxm5091 Glad to hear you got it working. Not that it matters that much now, I haven't updated the app since applying the patch but so far it's working fine.

**Comment by bbarthec** [comment]
I had a similar problem. I've tried several approaches (including usage of this patch provided in https://github.com/react-native-maps/react-native-maps/issues/4214#issuecomment-1132214457), but dangling markers were still there.

I've succeeded in solving my problem by removing usage of [`LayoutAnimation` from `react-native-reanimated`](https://docs.swmansion.com/react-native-reanimated/docs/api/LayoutAnimations/entryAnimations) completely from my app.

I realised this is the root cause, because this _dangling markers_ were only appearing on the map when I was navigating from other screen (I'm using `react-navigation/native-stack` BTW) that was using `FafeIn` or `FadeOut` anywhere in it's components hierarchy. This problem was not happening if I was navigating to my map screen from any screen that was not using `LayoutAnimations`.

My environment at the time the problem is solved:
```json
{
  "@react-navigation/bottom-tabs": "6.4.0",
  "@react-navigation/native-stack": "6.9.1",
  "react-native": "0.69.6",
  "react-native-maps": "1.3.2",
  "react-native-reanimated": "2.9.1",
}
```

**TL;DR** Complete removal of [`LayoutAnimation` from `react-native-reanimated`](https://docs.swmansion.com/react-native-reanimated/docs/api/LayoutAnimations/entryAnimations) solved the problem for me.

**Comment by KumarVijayendra** [comment]
We had similar issue and as pointed out by @bbarthec layout animations used are reason for it. 
We are using the patch shared by @l-barbosa mentioned in https://github.com/react-native-maps/react-native-maps/issues/4214#issuecomment-1132214457 and it has resolved issue for us. 

Patch is applied on react-native-reanimated@2.10.0

**Comment by MatthewLamperski** [comment]
> Sorry to hear that we're not dealing with the same case.
> 
> In case anyone is experiencing the same specific issue I was able to work around it by patching `react-native-reanimated`. You can find [the patch](https://github.com/l-barbosa/react-native-maps-reanimated-issue/blob/master/patches/react-native-reanimated%2B2.8.0.patch) in the demo repository.

I know this is a closed issue, but I just wanted to say thank you @l-barbosa. I'm the one that brought up the same issue in #4102. While working on an update for the same app that brought up the original problem, I ran into this problem yet again. And it is completely accurate that the issue is caused by reanimated - I was incorporating a layout animation. But I read through this _incredibly_ detailed issue, and applied the [patch](https://github.com/l-barbosa/react-native-maps-reanimated-issue/blob/master/patches/react-native-reanimated%2B2.8.0.patch), and now everything works great, even with the layout animations! Just wanted to revisit the issue and say thanks for getting to the root problem.

**Comment by sebastienlabine** [comment]
Thanks for the patch, however it would be nice to a long time solution for this bug as many people experience this bug. #4102

**Comment by Snoeprol** [comment]
Hi. Fairly new to react-native, but I changed the file `REAUIManager.mm` and applied the patch.
However, I still have the same issue. Is there anything else I need to do like reload the modules?

Best  to all :)!

**Comment by Snoeprol** [comment]
Hi. I have react-native-reanimated version `2.14.4` and I applied the patch.

I went to folder `/node_modules/` and added the necessary lines. This + rerunning the project using `npx expo start` should do the job right? Am I missing something? My markers are still not being removed. Did anyone find other workarounds?

Best, Mario

**Comment by Snoeprol** [comment]
Another solution I found is to give every Marker a unique `key` prop e.g.:
```react
    <Marker
      key={event.key}
...
   </Marker>
```

Hope this helps someone else who ignores warnings sometimes ;)!

**Comment by henrymoulton** [comment]
> Hi. I have react-native-reanimated version `2.14.4` and I applied the patch.
> 
> I went to folder `/node_modules/` and added the necessary lines. This + rerunning the project using `npx expo start` should do the job right? Am I missing something? My markers are still not being removed. Did anyone find other workarounds?
> 
> Best, Mario

Hi Mario, no you'll have to use `patch-package` to apply the patch: https://github.com/ds300/patch-package

**Comment by krakowski** [comment]
Stumbled upon the same issue on Android.

In my case the App gets really slow after using the `MapView` (zooming in & out, moving, etc.) for some time. Looking at the view hierarchy using [Flipper](https://github.com/facebook/flipper) revealed that `MapMarker` views stay in place even after removing them manually and triggering a hot-reload. In my case the default marker is not visible on the map, but the view hierarchy grows indefinitely and causes the whole App to slow down. The patch provided in https://github.com/react-native-maps/react-native-maps/issues/4214#issuecomment-1132214457 seems to be for iOS only.

Here's the most basic example I can think of triggering this behaviour.

```ts
{ locations.slice(0, 5).map((location, index) => (
    <Marker
      key={index}
      identifier={index.toString()}
      coordinate={{
          latitude: location.latitude,
          longitude: location.longitude,
      }}
      tracksViewChanges={false}
    />
  )
)}
```

`locations` contains some demo data (e.g `{ latitude: 45.234123, longitude: 17.938294}`). Changing the slices length after the initial render from `5` to `1` and going back to `5` again adds `4` new `MapMarker` views while keeping the old ones within the view hierarchy (total of `9` markers, while there should be only `5`). Here's a short recording showing the mentioned behavior in action.

![Peek 2023-11-17 20-02](https://github.com/react-native-maps/react-native-maps/assets/7255767/d90ed99a-de72-42ee-9a83-772a88eddaaf)


Does anyone know why this is happening or has any hints regarding this issue?
